src/vs/workbench/api/common/extHostTesting.ts

```diff
@@ -222,6 +228,110 @@ export class ExtHostTesting implements ExtHostTestingShape {
+		workspaceHierarchy.onDidChangeTest(node => {
+			const wrapper = TestItemFilteredWrapper.getWrapperForTestItem(node, document.uri);
+			if (wrapper.doesHaveNodeMatchingFilter(true)) {
+				onDidChangeTest.fire(wrapper);
+			} else {
+				onDidChangeTest.fire(wrapper.visibleParent);
+			}
+		});
+
+		return {
+			root: TestItemFilteredWrapper.getWrapperForTestItem(workspaceHierarchy.root, document.uri),
+			dispose: () => onDidChangeTest.dispose(),
+			onDidChangeTest: onDidChangeTest.event
+		};
+	}
+}
+
+/*
+ * A class which wraps a vscode.TestItem that provides the ability to filter a TestItem's children
+ * to only the children that are located in a certain vscode.Uri.
+ */
+class TestItemFilteredWrapper implements vscode.TestItem {
+	private static wrapperMap = new WeakMap<vscode.TestItem, TestItemFilteredWrapper>();
+
+	// Wraps the TestItem specified in a TestItemFilteredWrapper and pulls from a cache if it already exists.
+	public static getWrapperForTestItem(item: vscode.TestItem, filterToUri: vscode.Uri, parent?: TestItemFilteredWrapper) {
+		if (TestItemFilteredWrapper.wrapperMap.has(item)) {
+			return TestItemFilteredWrapper.wrapperMap.get(item)!;
+		}
+		const w = new TestItemFilteredWrapper(item, filterToUri, parent);
+		TestItemFilteredWrapper.wrapperMap.set(item, w);
+		return w;
+	}
+
+	public get label() {
+		return this.actual.label;
+	}
+
+	public get debuggable() {
+		return this.actual.debuggable;
+	}
+
+	public get description() {
+		return this.actual.description;
+	}
+
+	public get location() {
+		return this.actual.location;
+	}
+
+	public get runnable() {
+		return this.actual.runnable;
+	}
+
+	public get state() {
+		return this.actual.state;
+	}
+
+	public get children() {
+		// We only want children that match the filter.
+		return this.getWrappedChildren().filter(child => child.doesHaveNodeMatchingFilter());
+	}
+
+	public get visibleParent(): TestItemFilteredWrapper {
+		return this.doesHaveNodeMatchingFilter() ? this : this.parent!.visibleParent;
+	}
+
+	private readonly matchesUri: boolean;
+	private hasNodeMatchingFilter: boolean | undefined;
+
+	// Determines if the TestItem matches the filter. This would be true if:
+	// 1. We don't have a parent (because the root is the workspace root node)
+	// 2. The URI of the current node matches the filter URI
+	// 3. Some child of the current node matches the filter URI
+	public doesHaveNodeMatchingFilter(forceEvaluation?: boolean): boolean {
+		if (forceEvaluation || this.hasNodeMatchingFilter === undefined) {
+			this.hasNodeMatchingFilter = !this.parent
+				|| this.matchesUri
+				|| this.getWrappedChildren().some(child => child.doesHaveNodeMatchingFilter());
+		}
+
+		return this.hasNodeMatchingFilter;
+	}
+
+	constructor(public readonly actual: vscode.TestItem, private filterToUri: vscode.Uri, private readonly parent?: TestItemFilteredWrapper) {
+		this.matchesUri = this.actual.location?.uri.toString() === this.filterToUri.toString();
```

Comment:
Though unlikely, it's possible the test location will change and start matching the filter, so this should probably be evaluated at check time.